lib/documents.js

```diff
@@ -2716,6 +2716,253 @@ function urisOutputTransform(headers, data) {
+      throw new Error('Invalid value for outputStreamType. Value must be chunked or object.');
+    }
+    outputStream = jobState.jobOptions.outputStreamType === 'object'?
+        new stream.PassThrough({objectMode: true}): new stream.PassThrough();
+  } else {
+    outputStream = new stream.PassThrough({objectMode: true});
+  }
+  jobState.outputStream = outputStream;
+
+  let duplexStream = new stream.Duplex({
+    read(size){
+      return outputStream.read(size);
+    },
+    write(chunk, encoding, callback) {
+      inputStream.write(chunk, encoding, callback);
+    }
+  });
+
+  jobState.duplexStream = duplexStream;
+
+  if(jobState.jobOptions.concurrentRequests) {
+    const concurrentRequests = jobState.jobOptions.concurrentRequests;
+    if (concurrentRequests.multipleOf) {
+      if (!new Set(['forests', 'hosts']).has(String(concurrentRequests.multipleOf))) {
+        throw new Error('Invalid value for multipleOf. Value must be forests or hosts.');
+      }
+    } else {
+      concurrentRequests.multipleOf = 'forests';
+    }
+
+    if (concurrentRequests.multiplier && concurrentRequests.multiplier <= 0) {
+      throw new Error('concurrentRequests.multiplier cannot be less than one');
+    } else {
+      concurrentRequests.multiplier = 8;
+    }
+    jobState.jobOptions.concurrentRequests = concurrentRequests;
+  }
+  else {
+    jobState.jobOptions.concurrentRequests = {multipleOf:'forests', multiplier:8};
+  }
+  if(jobState.jobOptions.batchSize){
+    if(jobState.jobOptions.inputKind === 'array'){
+      throw new Error('batchSize not expected when inputKind is array.');
+    }
+    if(jobState.jobOptions.batchSize <=0){
+      jobState.jobOptions.batchSize = 250;
+    }
+  } else {
+    jobState.jobOptions.batchSize = 250;
+  }
+
+  requester.startRequest(operation).result(mlutil.callbackOn(jobState, onReadAllInit))
+      .catch((err) => operation.stream.emit('error', err));
+
+  return duplexStream;
+}
+
+function onReadAllInit(output) {
+  let maxRequesters = 0;
+  const jobState = this;
+
+  switch (jobState.jobOptions.concurrentRequests.multipleOf) {
+    case 'hosts':
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier *
+          new Set(output.map((forest) => forest.host)).size;
+      break;
+    default:
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier * output.length;
+      break;
+  }
+  jobState.startTime = Date.now();
+  for (let i = 0; i < maxRequesters; i++) {
+    jobState.requesterCount++;
+    onReadAllDocs(jobState,i);
+  }
+}
+
+function onReadAllDocs(jobState, readerId) {
+
+  if(jobState.error){
+    finishReader(jobState);
+    return;
+  }
+  const readBatchArray = [];
+  const batchSize = jobState.jobOptions.batchSize;
+  jobState.lastBatch = false;
+  jobState.retryCount = 100;
+
+  a:
+      for (let i = 0; i < batchSize; i++) {
+        const record = jobState.inputStream.read();
```

Comment:
It might be more readable for future maintenance to rename the record constant to uri because the string represents a document uri.